File: drills/ts/solutions/nc93_longest_common_subsequence.ts
export function longestCommonSubsequence(text1: string, text2: string): number {
  // Compare the last characters: equal means both are used and the answer is
  // one more than the rest, different means the best of dropping one or the
  // other. Filled row by row, only the previous row is ever needed.
  let previous = new Array<number>(text2.length + 1).fill(0);

  for (const a of text1) {
    const row = new Array<number>(text2.length + 1).fill(0);
    for (let j = 1; j <= text2.length; j++) {
      row[j] = a === text2[j - 1] ? previous[j - 1] + 1 : Math.max(previous[j], row[j - 1]);
    }
    previous = row;
  }

  return previous[text2.length];
}

File: drills/go/solutions/nc93_longest_common_subsequence.go
package main

func longestCommonSubsequence(text1 string, text2 string) int {
	// table[i][j]: LCS of text1[i:] and text2[j:]. Matching characters
	// extend the diagonal; otherwise take the better of dropping one.
	m, n := len(text1), len(text2)
	table := make([][]int, m+1)
	for i := range table {
		table[i] = make([]int, n+1)
	}
	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			if text1[i] == text2[j] {
				table[i][j] = 1 + table[i+1][j+1]
			} else {
				table[i][j] = max(table[i+1][j], table[i][j+1])
			}
		}
	}
	return table[0][0]
}
